scipy/stats/_morestats.py

```diff
@@ -1309,27 +1321,33 @@ def _all(x):
         message = ("The `optimizer` argument of `boxcox_normmax` must return "
                    "an object containing the optimal `lmbda` in attribute `x`.")
         raise ValueError(message)
-    else:
-        # Test if the optimal lambda causes overflow
+    elif ymax is None or not np.isinf(ymax):  # adjust the final lambda
         x = np.asarray(x)
-        max_x = np.max(x, axis=0)
-        istransinf = np.isinf(special.boxcox(max_x, res))
-        dtype = x.dtype if np.issubdtype(x.dtype, np.floating) else np.float64
-        if np.any(istransinf):
-            warnings.warn(
-                f"The optimal lambda is {res}, but the returned lambda is "
-                f"the constrained optimum to ensure that the maximum of the "
-                f"transformed data does not cause overflow in {dtype}.",
-                stacklevel=2
-            )
-
-            # Return the constrained lambda to ensure the transformation
-            # does not cause overflow
-            ymax = np.finfo(dtype).max / 100  # 100 is the safety factor
-            constrained_res = _boxcox_inv_lmbda(max_x, ymax)
+        xmax = np.max(x, axis=0)
+        ymax_res = special.boxcox(xmax, res)
+
+        if ymax is None:
+            # Test if the optimal lambda causes overflow
+            mask = np.isinf(ymax_res)
+            if np.any(mask):
+                dtype = x.dtype if np.issubdtype(x.dtype, np.floating) else np.float64
+                warnings.warn(
+                    f"The optimal lambda is {res}, but the returned lambda is "
+                    f"the constrained optimum to ensure that the maximum of the "
+                    f"transformed data does not cause overflow in {dtype}.",
+                    stacklevel=2
+                )
+
+                # Return the constrained lambda to ensure the transformation
+                # does not cause overflow
+                ymax = np.finfo(dtype).max / 100  # 100 is the safety factor
+        else:
+            mask = ymax_res > ymax
```

Comment:
I think it's reasonably to warn whenever `ymax_res > ymax`, not just when `ymax` is the default.

So at the top, you can do:
```python3
        if ymax is None:
            # Test if the optimal lambda causes overflow
            if np.any(mask):
                dtype = x.dtype if np.issubdtype(x.dtype, np.floating) else np.float64
            ymax = np.finfo(dtype).max / 100  # 100 is the safety factor
```

and from there on, the case of default `ymax` or user-specified `ymax` is treated the same way.
```python3
mask = ymax_res > ymax
if np.any(mask):
...
```

Reply:
Sure, but in this case we will compare `ymax_res` and `np.finfo(dtype).max / 100`, not check if `ymax_res` is inf.

I want use the following instead.
```python
        if ymax is None:
            # Test if the optimal lambda causes overflow
            mask = np.isinf(ymax_res)
            if np.any(mask):
                dtype = x.dtype if np.issubdtype(x.dtype, np.floating) else np.float64
                ymax = np.finfo(dtype).max / 100  # 100 is the safety factor
                end_msg = f"cause overflow in {dtype}."
        else:
            mask = ymax_res > ymax
            if np.any(mask):
                end_msg = "exceed specified `ymax`."

        if np.any(mask):
            message = (
                f"The optimal lambda is {res}, but the returned lambda is "
                f"the constrained optimum to ensure that the maximum of the "
                f"transformed data does not " + end_msg
            )
            warnings.warn(message, stacklevel=2)

Reply:
> Sure, but in this case we will compare

Yes. That was the intent.